botocore/utils.py

```diff
@@ -260,70 +265,108 @@ def __init__(self, timeout=DEFAULT_METADATA_SERVICE_TIMEOUT,
             proxies=get_environ_proxies(self._url),
         )
 
-    def _get_request(self, url, timeout, num_attempts=1):
+    def retrieve_iam_role_credentials(self):
+        try:
+            role_name = self._get_iam_role()
+            credentials = self._get_credentials(role_name)
+            if self._contains_all_credential_fields(credentials):
+                return {
+                    'role_name': role_name,
+                    'access_key': credentials['AccessKeyId'],
+                    'secret_key': credentials['SecretAccessKey'],
+                    'token': credentials['Token'],
+                    'expiry_time': credentials['Expiration'],
+                }
+            else:
+                # IMDS can return a 200 response that has a JSON formatted
+                # error message (i.e. if ec2 is not trusted entity for the
+                # attached role). We do not necessarily want to retry for
+                # these and we also do not necessarily want to raise a key
+                # error. So at least log the problematic response and return
+                # an empty dictionary to signal that it was not able to
+                # retrieve credentials.
+                logger.debug('Response received %s is missing '
+                             'credentials.', credentials)
+                return {}
+        except _RetriesExceededError:
+            logger.debug("Max number of attempts exceeded (%s) when "
+                         "attempting to retrieve data from metadata service.",
+                         self._num_attempts)
+        return {}
+
+    def _get_iam_role(self):
+        return self._get_request(
+            url=self._url,
+            num_attempts=self._num_attempts,
+            needs_retry=self._does_not_contain_role_name
+        ).text
+
+    def _get_credentials(self, role_name):
+        r = self._get_request(
+            url=self._url + role_name,
+            num_attempts=self._num_attempts,
+            needs_retry=self._does_not_contain_credentials
+        )
+        return json.loads(r.text)
+
+    def _get_request(self, url, num_attempts=1, needs_retry=None):
```

Comment:
Do we need to default to 1 here?  Seems like it's safer to just require you to provide `self._num_attempts` when calling this method.

Reply:
No we do not need a default. I'm fine with removing it.